### PR TITLE
Re-implement long polling with svar

### PR DESCRIPTION
**Pull request by rgrinberg**
The PR Does 3 things.

## Re-implement long polling with Svar

This is self explanatory as this was the only reason we introduced Svar's

## Fixes sync files on macos

This regression was introduced in https://github.com/ocaml/dune/pull/5135. We must fix it here to actually test the feature on macos. Unfortunately, fixing this uncovers a bug with the current diagnostics implementation. Which makes it impossible to move this into a separate PR (it's impossible to fix the issue with the old implementation).

## Remove status line output from the tests

Since status line output is now generated by reading from an Svar, it's not fully deterministic. It's totally redundant for testing though, so we can safely get rid of it.

### CODE REVIEW

File: bin/import.ml [hunk 1]
```diff
@@ -51,12 +51,20 @@ end = struct
...
+           | Build_finished_with_success_and_waiting ->
+             Pp.tag User_message.Style.Success (Pp.verbatim "Success")
+             ++ waiting
            | Build_finished_with_failure_and_waiting ->
-             Pp.nop
+             Pp.tag User_message.Style.Error (Pp.verbatim "Had errors")
+             ++ waiting
+           | Interrupting_current_build ->
+             Pp.tag User_message.Style.Error
+               (Pp.verbatim "Source files changed")
```
Comment (rgrinberg) [inline bin/import.ml:66]: @jeremiedimino it seems like the semantics of `Fiber.Svar.t` mean that some notifications might be missed. In particular, it's not a guarantee that the UI will observe the `Interrupting_current_build` event. In practice, I suppose it's not a big deal as users only care about the current results. However, I think they would prefer to always see the delimiters before/after builds.

Comment (jeremiedimino) [reply bin/import.ml:66]: We have to choose between a signal and an event semantic, and we choose a signal semantic, so it's expected we don't see all the state changes.

Why do you think users would want to see the delimiters? If it's to know whether a new build has started, then there is a simple solution for that: we can include the build number in the state.

Comment (jeremiedimino) [reply bin/import.ml:66]: Ah wait, I didn't realise this comment was for the console printer. For the console, I agree that we want to see the new build delimiters, given that it is a continuous output. I remember thinking about this when I did this change, and I convinced myself that this was still the case because we call `Console.Status_line.refresh` on every iteration (via the `Tick` event handler).

Comment (jeremiedimino) [reply bin/import.ml:66]: But now that I write this down, it feels a bit fragile. Maybe we should add a primitive `Fiber.Svar.changes : 'a t -> 'a Fiber.Stream.In.t`?

Comment (jeremiedimino) [reply bin/import.ml:66]: Scratch that, looking again at the code, it looks all good to me because the `*** New build ***` messages are printed separately. The printing code I moved was all for temporary messages.

File: src/dune_rpc_server/dune_rpc_server.mli [hunk 2]
```diff
@@ -111,12 +111,22 @@ module Handler : sig
...
-       's t
-    -> 'a Procedures.Poll.t
-    -> on_poll:('s Session.t -> Poller.t -> 'a option Fiber.t)
-    -> on_cancel:('s Session.t -> Poller.t -> unit Fiber.t)
+  val implement_long_poll :
+       _ t
+    -> 'diff Procedures.Poll.t
+    -> 'state Fiber.Svar.t
+    -> equal:('state -> 'state -> bool)
+    -> make_diff:(last:'state option -> now:'state -> 'diff)
```
Comment (jeremiedimino) [inline src/dune_rpc_server/dune_rpc_server.mli:119]: Personally, I'd just call this `diff`:

```suggestion
    -> diff:(last:'state option -> now:'state -> 'diff)
```

**Review by jeremiedimino — APPROVED** [review]
Looks good!

File: bin/import.ml (hunk 1 above)
Comment (jeremiedimino) [reply bin/import.ml:66]: Hmm, some tests are failing. I can see the issue: `Console.Status_line.set (Constant _)` always print something, but `Console.Status_line.set (Live _)` prints nothing if the output is not a terminal

Comment (jeremiedimino) [reply bin/import.ml:66]: It seems complicated to fix this. I liked this refactoring as it centralised the logic to render the status line and removed some duplication between the build system state and scheduler events, but it seems complicated to get back the old behaviour. For now you should just drop my commit.

**Comment by rgrinberg** [comment]
@snowleopard this PR is ready. Jeremie already reviewed but it wouldn't hurt to get a second opinion.

**Comment by snowleopard** [comment]
> @snowleopard this PR is ready. Jeremie already reviewed but it wouldn't hurt to get a second opinion.

Sure, I'll try to have a look soon!

**Comment by rgrinberg** [comment]
Ping @snowleopard

File: otherlibs/dune-rpc-lwt/test/dune_rpc_lwt_tests.ml [hunk 3]
```diff
@@ -77,7 +72,7 @@ let%expect_test "run and connect" =
   Lwt_main.run
     (let* root_dir = Lwt_io.create_temp_dir () in
      Sys.chdir root_dir;
-     let build = build_watch ~root_dir ~suppress_stderr:false in
+     let build = build_watch ~root_dir in
```
Comment (snowleopard) [inline otherlibs/dune-rpc-lwt/test/dune_rpc_lwt_tests.ml:75]: There is no PR description but from the title I suppose the intent here is only to change the implementation. It seems strange, therefore, that here we're dropping the `suppress_stderr` argument.

**Comment by snowleopard** [comment]
> Ping @snowleopard

Looking. Sorry for the delay -- we're in the middle of an internal build systems crisis :)

**Comment by snowleopard** [comment]
Btw, could you add a brief description to the PR? It seems like it's changing some behaviour, so it would be great to know what's intended.

File: src/dune_engine/build_system.mli [hunk 4]
```diff
@@ -68,17 +68,27 @@ module Progress : sig
...
 
   val is_determined : t -> bool
 end
 
-val get_current_progress : unit -> Progress.t
+module State : sig
+  type t =
+    | Initializing
+    | Building of Progress.t
+    | Interrupting_current_build
```
Comment (snowleopard) [inline src/dune_engine/build_system.mli:84]:
```suggestion
    | Restarting_current_build
```
(I think "restarting" is much clearer than "interrupting".)

File: src/dune_engine/build_system.mli [hunk 5]
```diff
@@ -68,17 +68,27 @@ module Progress : sig
...
+    | Build_finished_with_success_and_waiting
+    | Build_finished_with_failure_and_waiting
```
Comment (snowleopard) [inline src/dune_engine/build_system.mli:86]:
```suggestion
    | Restarting_current_build
    | Build_succeeded__now_waiting_for_changes
    | Build_failed__now_waiting_for_changes
```
I think "restarting" is much clearer than "interrupting". Also tried to make the two other variants a bit clearer (not 100% sure that I succeeded).

File: src/dune_engine/build_config.mli [hunk 6]
```diff
@@ -56,6 +56,22 @@ module Error : sig
...
+  module Event : sig
+    type nonrec t =
+      | Add of t
+      | Remove of t
+  end
+
+  module Set : sig
+    type nonrec t =
+      { current : t Id.Map.t
+      ; stamp : int
```
Comment (snowleopard) [inline src/dune_engine/build_config.mli:68]: Could you add a comment explaining the purpose of the `stampt`?

File: src/dune_engine/build_config.mli [hunk 7]
```diff
@@ -56,6 +56,22 @@ module Error : sig
...
+      ; last_event : Event.t option
```
Comment (snowleopard) [inline src/dune_engine/build_config.mli:69]: Same for `last_event`.

File: src/dune_engine/build_system.ml [hunk 8]
```diff
@@ -30,20 +30,100 @@ end = struct
...
+
+  let equal { number_of_rules_discovered; number_of_rules_executed } t =
+    Int.equal number_of_rules_discovered t.number_of_rules_discovered
+    && Int.equal number_of_rules_executed t.number_of_rules_executed
+
+  let complete t = t.number_of_rules_executed
+
+  let remaining t = t.number_of_rules_discovered - t.number_of_rules_executed
+
+  let is_determined { number_of_rules_discovered; number_of_rules_executed } =
```
Comment (snowleopard) [inline src/dune_engine/build_system.ml:48]: Not really related to this PR but I have trouble understanding what `is_determined` means. I think `number_of_rules_executed <> 0` implies `number_of_rules_discovered <> 0`, so this expression can be simplified. If that's true, can we just expose `number_of_rules_discovered`, whose meaning is much clearer than `is_determined`?

File: src/dune_engine/build_system.ml [hunk 9]
```diff
@@ -30,20 +30,100 @@ end = struct
...
+  let add_error error =
+    let set =
+      let set = Svar.read errors in
+      let current = Error.Id.Map.set set.current (Error.id error) error in
+      { Error.Set.current
+      ; stamp = set.stamp + 1
+      ; last_event = Some (Add error)
+      }
+    in
+    Svar.write errors set
```
Comment (snowleopard) [inline src/dune_engine/build_system.ml:126]: I'm a bit nervous about such non-atomic read-then-write state modifications. It feels like we should have an atomic `update` function to make this code more robust (it's probably fine right now but who knows how we refactor things in future).

File: src/dune_engine/build_system.ml [hunk 10]
```diff
@@ -227,7 +305,12 @@ end = struct
     Option.iter t.stats ~f:(fun stats ->
         let module Event = Chrome_trace.Event in
         let event =
-          let args = [ ("value", `Int !State.rule_total) ] in
+          let rule_total =
+            match Fiber.Svar.read State.t with
+            | Building progress -> progress.number_of_rules_discovered
+            | _ -> assert false
```
Comment (snowleopard) [inline src/dune_engine/build_system.ml:311]: Worth renaming the function to `report_evaluated_rule_exn` then.

File: src/dune_file_watcher/dune_file_watcher.ml [hunk 11]
```diff
@@ -245,33 +245,58 @@ module Buffer = struct
...
-
-let consume_sync_event table path =
-  let basename = Filename.basename path in
-  match Table.find table basename with
-  | None -> None
-  | Some id ->
-    Fpath.unlink_no_err path;
-    Table.remove table basename;
-    Some id
+module Fs_sync : sig
```
Comment (snowleopard) [inline src/dune_file_watcher/dune_file_watcher.ml:248]: I'm confused by changes in this module. They seem to have nothing to do with `Svar`s. Shall this be moved to a separate PR?

File: src/dune_engine/build_system.ml (hunk 8 above)
Comment (snowleopard) [reply src/dune_engine/build_system.ml:48]: (It's fine to just turn this into a `CR-someday` comment if you'd rather not change anything here.)

File: src/dune_rpc_impl/server.ml [hunk 12]
```diff
@@ -146,12 +143,57 @@ let handler (t : t Fdecl.t) : 'a Dune_rpc_server.Handler.t =
...
+  let () =
+    let diff ~last:_ ~(now : Build_system.State.t) =
+      match now with
+      | Initializing -> Progress.Waiting
+      | Interrupting_current_build -> Interrupted
+      | Build_finished_with_success_and_waiting -> Success
+      | Build_finished_with_failure_and_waiting -> Failed
+      | Building now ->
+        let remaining =
+          now.number_of_rules_discovered - now.number_of_rules_executed
```
Comment (snowleopard) [inline src/dune_rpc_impl/server.ml:189]: It's a bit weird that we have some code duplication here (with `build_system.ml`).

**Review by snowleopard — CHANGES_REQUESTED** [review]
Generally I like the changes, thanks!

I had a less careful look at `src/dune_rpc_server/dune_rpc_server.ml` but I hope the existing tests cover the behaviour.

I left quite a few minor comments/suggestions, please have a look.

Also, this PR appears to be doing more than its title suggests. Could you add a description to document all intended changes and their motivations?

File: src/dune_engine/build_system.mli (hunk 5 above)
Comment (rgrinberg) [reply src/dune_engine/build_system.mli:86]: The names are fine but why double underscores? Those are already used for name mangling

File: src/dune_file_watcher/dune_file_watcher.ml (hunk 11 above)
Comment (rgrinberg) [reply src/dune_file_watcher/dune_file_watcher.ml:248]: I tried doing that but with no success (https://github.com/ocaml/dune/pull/5363). We need this commit for this PR to be properly tested, but rpc itself is a little buggy and needs the fixes in this PR. There's an individual commit for it but I can include it in the PR description as well.

File: src/dune_engine/build_system.ml (hunk 8 above)
Comment (rgrinberg) [reply src/dune_engine/build_system.ml:48]: This function is unused. I removed it.

File: src/dune_engine/build_system.ml (hunk 9 above)
Comment (rgrinberg) [reply src/dune_engine/build_system.ml:126]: We rely on code outside of fiber binds to be atomic just about everywhere. So I'm not sure there's anything particularly special here.

File: src/dune_rpc_impl/server.ml (hunk 12 above)
Comment (rgrinberg) [reply src/dune_rpc_impl/server.ml:189]: Do you mean with the type `Build_system.State.t`? Yes it's a bit annoying but the analogous type in RPC has stability guarantees that this type does not.

File: src/dune_engine/build_config.mli (hunk 7 above)
Comment (rgrinberg) [reply src/dune_engine/build_config.mli:69]: I moved it all to Build_system

**Comment by rgrinberg** [comment]
> Also, this PR appears to be doing more than its title suggests. Could you add a description to document all intended changes and their motivations?

Done

File: src/dune_engine/build_config.mli (hunk 6 above)
Comment (rgrinberg) [reply src/dune_engine/build_config.mli]: I improve this module a bunch. Now all of this stamp stuff is completely private. So there's no need to explain this somewhat error prone and confusing stuff.

File: otherlibs/dune-rpc-lwt/test/dune_rpc_lwt_tests.ml (hunk 3 above)
Comment (snowleopard) [reply otherlibs/dune-rpc-lwt/test/dune_rpc_lwt_tests.ml:75]: Sounds good.

File: src/dune_engine/build_system.mli (hunk 5 above)
Comment (snowleopard) [reply src/dune_engine/build_system.mli]: As we discussed offline, double underscores here are meant to represent commas, as in "build failed, now waiting for changes". 

Looks like `Interrupting_current_build` isn't renamed to `Restarting_current_build` yet.

File: src/dune_engine/build_config.mli (hunk 6 above)
Comment (snowleopard) [reply src/dune_engine/build_config.mli]: Awesome!

File: src/dune_engine/build_system.ml (hunk 9 above)
Comment (snowleopard) [reply src/dune_engine/build_system.ml:225]: OK, sounds good.

File: src/dune_file_watcher/dune_file_watcher.ml (hunk 11 above)
Comment (snowleopard) [reply src/dune_file_watcher/dune_file_watcher.ml:248]: Got it.

File: src/dune_engine/build_system.mli (hunk 5 above)
Comment (rgrinberg) [reply src/dune_engine/build_system.mli]: Added the missing rename.

File: src/dune_engine/build_system.ml [hunk 13]
```diff
@@ -29,21 +29,200 @@ end = struct
...
+    match current with
+    | Building current ->
+      Svar.write t
+        (Building
+           { current with
+             number_of_rules_executed = current.number_of_rules_executed + 1
+           })
+    | _ -> assert false
+
+  let start_rule () =
```
Comment (snowleopard) [inline src/dune_engine/build_system.ml:205]: One more candidate for `_exn`.

**Review by snowleopard — APPROVED** [review]
Looks great, thanks!

File: src/dune_engine/build_system.ml [hunk 14]
```diff
@@ -372,7 +554,7 @@ end = struct
     let { Rule.id = _; targets; dir; context; mode; action; info = _; loc } =
       rule
     in
-    start_rule rule;
+    let* () = Memo.Build.of_non_reproducible_fiber (State.start_rule_exn ()) in
```
Comment (snowleopard) [inline src/dune_engine/build_system.ml:557]: @rgrinberg I found the regression ^ :)

Comment (snowleopard) [reply src/dune_engine/build_system.ml:557]: (Fix coming)
